content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md

```diff
@@ -0,0 +1,383 @@
+
+```cpp
+AZStd::vector<AZStd::string> componentNames;
+AzToolsFramework::EditorComponentAPIBus::BroadcastResult(
+    componentNames,
+    &AzToolsFramework::EditorComponentAPIRequests::FindComponentTypeNames,
+    typeIds
+);
+```
+
+{{ note }}
+You can query a list of services that components provide by calling the component's `GetProvidedServices()` method.
+{{ /note }}
+
+### Add buttons
+
+Next, you will loop through the list of components, create buttons, and add them to the `gridLayout`. Refer to the following snippet in `ShapeEXampleQWidget.cpp`.
+
+For each component:
+
+1. Store the component's name and type from the list of component names and types that you queried earlier.
+
+2. Create a button using the QButton object called `shapeButton`. 
+   - Pass in `name` to add the component's name onto the button. 
+   - Set the button minimum height to `40` by using `setMinimumHeight(...)`. 
+
+3. Later, you will connect `shapeButton` to a signal listener. 
+
+4. Add the button to `gridLayout`, which is split into three columns. 
+
+```cpp
+const int maxColumnCount = 3;
+for (int i = 0; i < componentNames.size(); ++i)
+{
+    AZStd::string name = componentNames[i];
+    AZ::TypeId typeId = typeIds[i];
+
+    QPushButton* shapeButton = new QPushButton(QString::fromUtf8(name.c_str()), this);
+    shapeButton->setMinimumHeight(40);
+
+    // Connect 
+
+    int row = i / maxColumnCount;
+    int column = i % maxColumnCount;
+    gridLayout->addWidget(shapeButton, row, column);
+}
+
+```
+
+### Add a signal listener with a lambda handler
+
+Next, you will create a signal listener such that when the user clicks a button, an entity with the corresponding Shape component will spawn in your scene. 
+
+In this example, the signal listener uses a lambda handler, so you can define the C++ function without defining a slot.
+
+1. Define `CreateEntityWithShapeComponent` as a standard C++ function in `ShapeExampleWidget.h`. This function accepts an `AZ::TypeId&` as a parameter so it can create an entity with the specified component type.
+
+    ```cpp
+    private:
+        void CreateEntityWithShapeComponent(const AZ::TypeId& typeId);
+    ```
+
+2. In the for loop that you created earlier, create a connection from the `clicked` signal in the `shapeButton` to the lambda function in `this`. The lambda function calls `CreateEntityWithShapeComponent(...)` and passes in the `typeId`. 
+
+    ```cpp
+    QObject::connect(shapeButton, &QPushButton::clicked, this, [this, typeId]() {
+        CreateEntityWithShapeComponent(typeId);
+    });
+    ```
+
+
+
+### Communicate with EBusses
+
+Now, you will define `CreateEntityWithShapeComponent(...)`, which uses communicates with O3DE EBusses to create a new entity with a component specified by the `typeId` parameter.
+
+1. Use the namespace `AzToolsFramework`. This namespace provides APIs to the necessary Editor EBusses. 
```

Comment:
```suggestion
1. Use the namespace `AzToolsFramework`. This namespace provides APIs to the necessary Editor EBuses. 
```

I'm not sure if we want to call this out explicitly. I did the `using namespace AzToolsFramework` just to make some of the ebus calls shorter/more readable, since otherwise you'd have to put `AzToolsFramework::` on the beginning of each of them. So maybe we could just say that?  Or nothing at all